### PR TITLE
opt: add implicit FOR UPDATE locks for generic query plans

### LINKED ISSUE #137352: opt: apply implicit `SELECT FOR UPDATE` to generic query plans
In a generic query plan, `FOR UPDATE` locks are not currently applied to reads of mutation statements. This can significantly increase contention in some workloads, hurting overall performance. We should apply `FOR UPDATE` locks to generic query plans in the same way that we do for non-generic query plans.

Jira issue: CRDB-45538

### PR DESCRIPTION
**Pull request by mgartner**
Implicit `FOR UPDATE` locks are now added to the reads in
`INSERT .. ON CONFLICT`, `UPSERT`, `UPDATE`, and `DELETE` queries when
those mutations use generic query plans.

Fixes #137352

Release note: None

### CODE REVIEW

**Comment by mgartner** [comment]
I added the tests for `DELETE`, but figured I'd wait until #137069 is merged to add support for generic query plans.

**Review by rytaft — DISMISSED** [review]
Nice!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-:-ODwkUmA0jvN3W9nlOvb:b-83sjoh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwkJ4194s3ZC8XyNSi:-ODwkJ4194s3ZC8XyNSj:bqxoby7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1203)):*
> ```Go
> 		// same as the table being updated. Also, don't lock rows if there is an
> 		// ON condition so that we don't lock rows that won't be updated.
> 		if updateTableStableID == lookupTableStableID && t.On.IsTrue() {
> ```

Can you add a test to show that we don't apply the lock if ON is not true?
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwjvJy79aISvo2su-v:-ODwjvJy79aISvo2su-w:bpb0r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1237)):*
> ```Go
> 	case *memo.LookupJoinExpr:
> 		input = join.Input
> 		if inner, ok := input.(*memo.LookupJoinExpr); ok && inner.Table == join.Table {
> ```

nit: add a comment to explain this

**Review by michae2 — COMMENTED** [review]
Nice fix!

> Reviewed 2 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-:-ODwk_WlBKxKM97Lds8X:b-k6moza)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwfsXS94AjP5lLFSq7:-ODwfsXS94AjP5lLFSq8:b6au2av):*
Do INSERT statements use implicit for update locks? I didn't think they did, but maybe they do for ON CONFLICT?
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwgIb_3SdWYKKnDL-e:-ODwgIb_3SdWYKKnDL-f:b-odbur9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1196)):*
> ```Go
> 	case *memo.ScanExpr:
> 		return t.Table
> 	case *memo.LookupJoinExpr:
> ```

This will also add implicit locking in custom plans for statements like the following, which did not have it before:

```sql
CREATE TABLE ab (a INT PRIMARY KEY, b INT);
CREATE TABLE cd (c INT PRIMARY KEY, d INT);
UPDATE ab SET b = b + 1 FROM cd WHERE a = d AND c = 1;
UPDATE ab AS ab1 SET b = ab1.b + 1 FROM ab AS ab2 WHERE ab1.a = ab2.b + 1 AND ab2.a = 1;
```

I think this is fine for the first statement, which will only lock rows that are modified, but I'm not sure about the second statement. The lock on the initial scan of ab2 could lock a row is not going to be updated.

**Review by michae2 — APPROVED** [review]
:lgtm:

> Reviewed 2 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-:-ODwlQZE0x_3XfLVq-K0:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwgIb_3SdWYKKnDL-e:-ODwkyeFB-HGFVl8JZGS:b-1sk6ab) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1196)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This will also add implicit locking in custom plans for statements like the following, which did not have it before:

```sql
CREATE TABLE ab (a INT PRIMARY KEY, b INT);
CREATE TABLE cd (c INT PRIMARY KEY, d INT);
UPDATE ab SET b = b + 1 FROM cd WHERE a = d AND c = 1;
UPDATE ab AS ab1 SET b = ab1.b + 1 FROM ab AS ab2 WHERE ab1.a = ab2.b + 1 AND ab2.a = 1;
```

I think this is fine for the first statement, which will only lock rows that are modified, but I'm not sure about the second statement. The lock on the initial scan of ab2 could lock a row is not going to be updated.
</blockquote></details>

Oh, I take it back, the second statement won't add a lock on the scan of ab2. So never mind, this seems good.

**Review by yuzefovich — COMMENTED** [review]
Nice!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-:-ODwwwvRDKtkpuVV1skk:b-83sjoh)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwfsXS94AjP5lLFSq7:-ODwtwCM82BluIBoJFNz:b-1zcxue):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do INSERT statements use implicit for update locks? I didn't think they did, but maybe they do for ON CONFLICT?
</blockquote></details>

No, I don't think we use implicit SFU for any type of INSERT.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwvH2Q2RUL321gvst_:-ODwvH2Q2RUL321gvsta:byrr9x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1184)):*
> ```Go
> 	}
> 
> 	// Try to match the Update's input expression against the pattern:
> ```

nit: consider updating this comment to match the new pattern.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODww8cI2Cerh9e92UMD:-ODww8cI2Cerh9e92UME:b-u5q1e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1204)):*
> ```Go
> 		// ON condition so that we don't lock rows that won't be updated.
> 		if updateTableStableID == lookupTableStableID && t.On.IsTrue() {
> 			return t.Table
> ```

I wonder whether we should be a bit more restrictive here, i.e. should we assert a particular shape of the input to `LookupJoinExpr` (for example, that it's a `ValuesExpr` or another `LookupJoinExpr` followed by `ValuesExpr`)? I can't think of case where the current code could lead to locking unnecessary rows, but we seem to have been more prescriptive in the existing cases where we apply implicit SFU.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137360)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwfsXS94AjP5lLFSq7:-ODwmIYX6gyxyRgwUrzZ:bt2vr1b):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

No, I don't think we use implicit SFU for any type of INSERT.
</blockquote></details>

Fixed.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwvH2Q2RUL321gvst_:-OE1634PAjJotBKgfM1_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1184)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider updating this comment to match the new pattern.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwgIb_3SdWYKKnDL-e:-OE1645TAnUhnMa6MqdJ:b-vkx6w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1196)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Oh, I take it back, the second statement won't add a lock on the scan of ab2. So never mind, this seems good.
</blockquote></details>

I added a test for the second case with a TODO because we don't use _any_ locks in that case.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1203 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwkJ4194s3ZC8XyNSi:-ODwq0Hm95gpN30IVfZO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1203)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you add a test to show that we don't apply the lock if ON is not true?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1204 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODww8cI2Cerh9e92UMD:-OE173nLDcfpvmE_y7QS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1204)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I wonder whether we should be a bit more restrictive here, i.e. should we assert a particular shape of the input to `LookupJoinExpr` (for example, that it's a `ValuesExpr` or another `LookupJoinExpr` followed by `ValuesExpr`)? I can't think of case where the current code could lead to locking unnecessary rows, but we seem to have been more prescriptive in the existing cases where we apply implicit SFU.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1237 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-ODwjvJy79aISvo2su-v:-OE16Csa0kuKwhCBOeWI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee65cc3d4fe5131874bde3c3949720b0db42af01/pkg/sql/opt/exec/execbuilder/mutation.go#L1237)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a comment to explain this
</blockquote></details>

Done.

**Review by yuzefovich — APPROVED** [review]
Nice! :lgtm:

> Reviewed 2 of 5 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137360#-:-OE1YeI_5rHGVHVrZS2P:be8rotp)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rytaft)

**Comment by mgartner** [comment]
TFTRs!

bors r+

**Comment by yuzefovich** [comment]
(Becca's review had the "requested changes" status which was blocking the merge.)

bors r+

**Comment by mgartner** [comment]
>(Becca's review had the "requested changes" status which was blocking the merge.)

Ah, thanks!

**Pull request merged**
